File: 039CombinationSum/CombinationSum.py
# -*- coding: utf-8 -*-

from typing import List

class CombinationSum:
    def combinationSum(
        self,
        candidates: "List[int]",
        target: "int"
    ) -> "List[List[int]]":
        ans = []
        self.__combSum_(ans, candidates, [], target, 0)
        return ans
    
    def __combSum_(
        self,
        ans: "List[List[int]]",
        candidates: "List[int]",
        current: "List[int]",
        remain: "int",
        start: "int"
    ) -> "None":
        if remain < 0: return
        elif remain == 0: ans.append(current[:])
        else:
            for i in range(start, len(candidates)):
                current.append(candidates[i])
                self.combSum_(
                    ans,
                    candidates,
                    current,
                    remain - candidates[i],
                    i
                )
                current.pop()


def main():
    test = CombinationSum()
    print(test.combinationSum([2, 3, 6, 7], 7))

if __name__ == "__main__":
    main()

File: 039CombinationSum/CombinationSum.cpp


#include<vector>

class CombinationSum {
public:
    std::vector<std::vector<int> > combinationSum(
        std::vector<int>& candidates, int target
    ) {
        std::vector<std::vector<int> > ans;
        std::vector<int> cur;
        this->combSumHelper(candidates, target, ans, cur, 0);
        return ans;
    }

private:
    void combSumHelper(
        std::vector<int>& nums,
        int target,
        std::vector<std::vector<int> >& ans,
        std::vector<int>& cur,
        int idx
    ) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            ans.push_back(cur);
            return;
        }
        for (int i = idx; i < nums.size(); ++i) {
            cur.push_back(nums[i]);
            this->combSumHelper(nums, target - nums[i], ans, cur, i);
            cur.pop_back();
        }
    }
};